host/src/attribute_server.rs

```diff
@@ -10,26 +10,51 @@ use crate::codec;
 use crate::cursor::WriteCursor;
 use crate::prelude::Connection;
 use crate::types::uuid::Uuid;
-
-#[derive(Debug, PartialEq)]
-pub enum WorkResult {
-    DidWork,
-    GotDisconnected,
-}
+use crate::Error;
 
 const MAX_NOTIFICATIONS: usize = 4;
-pub struct NotificationTable<const ENTRIES: usize> {
+pub(crate) struct NotificationTable<const ENTRIES: usize> {
     state: [(u16, ConnHandle); ENTRIES],
 }
 
-pub struct AttributeServer<'d, M: RawMutex, const MAX: usize> {
-    pub(crate) table: AttributeTable<'d, M, MAX>,
+/// A GATT server capable of processing the GATT protocol using the provided table of attributes.
+pub struct AttributeServer<'values, M: RawMutex, const MAX: usize> {
+    pub(crate) table: AttributeTable<'values, M, MAX>,
     pub(crate) notification: Mutex<M, RefCell<NotificationTable<MAX_NOTIFICATIONS>>>,
 }
 
-impl<'d, M: RawMutex, const MAX: usize> AttributeServer<'d, M, MAX> {
+pub(crate) mod sealed {
+    use super::*;
+
+    pub trait DynamicAttributeServer {
+        fn process(&self, connection: &Connection, packet: &AttReq, rx: &mut [u8]) -> Result<Option<usize>, Error>;
+        fn should_notify(&self, connection: &Connection, cccd_handle: u16) -> bool;
+        fn set(&self, characteristic: u16, input: &[u8]) -> Result<(), Error>;
+    }
+}
+
+/// Type erased attribute server
+pub trait DynamicAttributeServer: sealed::DynamicAttributeServer {}
```

Comment:
It is to prevent a public trait methods from being called outside the crate. It's used a lot in the embassy HALs for instance, so that's where I learned it.

Reply:
fyi you can replace `mod sealed` with `#[allow(private_bounds)]` which is a little less boilerplatey (like in embassy-rs/embassy#2730)

I always thought the reason was to stop external trait _`impl`s_ though. guess it’s the same thing really- can’t implement or call a method if you can’t see it.